pkg/apis/routing/v1alpha1/inference_router.go

```diff
@@ -0,0 +1,111 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+)
+
+// InferenceRouterSpec is the top level type for this resource
+// A router contains a set of strategies
+type InferenceRouterSpec struct {
+	// Routes is a list of route which can receive traffic
+	// All routes are expected to have an equivalent data plane interface
+	// RouteSpec is validated based on RouterType
+	// +required
+	Routes []RouteSpec `json:"routes"`
+	// Splitter splits the traffic among multiple routes, the number of routes should be equal to the
+	// the elements in weight array and the weights should sum to 100.
+	// +optional
+	Splitter *SplitterSpec `json:"splitter,omitempty"`
+	// AB testing between new and old model and compare the performance based on predefined metric.
+	// +optional
+	ABTest *ABTestSpec `json:"abTest,omitempty"`
+	// MultiArmBandit is an inference router node which routes traffic to the best model most of time,
+	// it also explores the other models, e.g epsilon greedy strategy routes to a random model with
+	// probability e and best model with probability 1-e.
+	// +optional
+	MultiArmBandit *MultiArmBanditSpec `json:"multiArmBandit,omitempty"`
+	// Ensemble is an inference router node which can combine results from previous steps
+	// +optional
+	Ensemble *EnsembleSpec `json:"ensemble,omitempty"`
+	// Pipeline defines a series of steps which are depending on previous step.
+	// Only one root node is allowed and all other routes should specify dependencies with the consumes field.
+	// The graph should form a valid DAG, no cycles are allowed, branches are allowed but they needs
+	// be mutually exclusive. Each route on the pipeline can be either an InferenceService or an Inference Router.
+	// +optional
+	Pipeline *PipelineSpec `json:"pipeline,omitempty"`
+}
+
+// RouteSpec defines the available routes in this router. Route functions reference routes by Name
+type RouteSpec struct {
+	// The name for the route
+	Name string `json:"name"`
+	// The destination of the route
+	Destination duckv1.Destination `json:",inline"`
+	// The header keys must be lowercase and use hyphen as the separator,
+	// e.g. _x-request-id_.
+	//
+	// Header values are case-sensitive and formatted as follows:
+	// - `exact: "value"` for exact string match
+	// - `prefix: "value"` for prefix-based match
+	// - `regex: "value"` for regex-based match
+	Headers map[string]*StringMatch `json:"headers,omitempty"`
+	// This is a list of pipeline route names that should be executed before the current route.
+	// The field is only valid for pipeline routing type
+	Consumes []string `json:"consumes,omitempty"`
```

Comment:
Not sure I like the Consumes which is only used in Pipeline. Seems confusing. Maybe can rethink later?

Reply:
sure, open for ideas. The advantage is to reuse the same `RouteSpec` and not introducing a new pipeline CRD, or you can think of other routes types all have an implicit `consumes` for the root.

Reply:
I might simply remove Consumes, for now. I think the whole model of defining dependencies might be the wrong direction. It results in a really challenging to understand semantic. You can build dags as follows:
```
Pipeline1 { Pipeline 2, Pipeline 3}
Pipeline2 { Transformer1, InferenceService1 }
Pipeline3{ Transformer1, InferenceService2 }
```